sklearn/tree/_criterion.pyx

```diff
@@ -676,6 +676,72 @@ cdef class Gini(ClassificationCriterion):
         impurity_right[0] = gini_right / self.n_outputs
 
 
+cdef class HellingerDistance(ClassificationCriterion):
+    r"""Hellinger distance criterion.
+
+    This handles cases where the target is a classification taking values 0 and 1.
+    If k represents a node before split, k1 and k2 represent the children nodes after split,
+    left_0 represents observations of class 0 in node k1, right_0 represents observations of class 0 in node k2,
+    left_1 represents observations of class 1 in node k1, right_1 represents observations of class 1 in node k2,
+    parent_0 represents observations of class 0 in node k, parent_1 represents observations of class 1 in node k,
+    then let
+
+        count_k1 = (\sqrt{\frac{N_{left_0}}{N_{parent_0}}}-\sqrt{\frac{N_{left_1}}{N_{parent_1}}})^2
+
+    be the partial hellinger distance score for the k1 node.
+
+        count_k2 = (\sqrt{\frac{N_{right_0}}{N_{parent_0}}}-\sqrt{\frac{N_{right_1}}{N_{parent_1}}})^2
+
+    be the partial hellinger distance score for the k2 node.
+
+    Hellinger distance score is then defined as:
+
+        hellinger_distance = \sqrt{count_k1+count_k2}
+    """
+    cdef double node_impurity(self) nogil:
+        cdef:
+            double impurity_left
+            double impurity_right
+
+        self.children_impurity(&impurity_left, &impurity_right)
+
+        return impurity_right + impurity_left
+
+    cdef void children_impurity(self, double* impurity_left,
+                                double* impurity_right) nogil:
+        cdef:
+            SIZE_t* n_classes = self.n_classes
+            double* sum_left = self.sum_left
+            double* sum_right = self.sum_right
+            double hellinger_left = 0.0
+            double hellinger_right = 0.0
+            double count_k1 = 0.0
+            double count_k2 = 0.0
+            SIZE_t k, c
+
+        for k in range(self.n_outputs):
+            count_k1 = 0.0
+            count_k2 = 0.0
+            if(sum_left[0] + sum_right[0] > 0):
+                count_k1 = sqrt(sum_left[0] / (sum_left[0] + sum_right[0]))
+            if(sum_left[1] + sum_right[1] > 0):
+                count_k2 = sqrt(sum_left[1] / (sum_left[1] + sum_right[1]))
+
+            hellinger_left += 1 - pow((count_k1  - count_k2), 2)
```

Comment:
@jmloyola I agree its not according to the formula, here is why

I've added the "1-", which is not in hellinger formula, in order to adapt the criterion to sklearn tree split design
The splitter in sklearn is searching for the best split, by looking for the maximum drop in split criterion score between parent and children nodes
However Hellinger distance criterion score acts opposite to Gini and Entropy

Gini - the lower score, the better split. Perfect split gini=0. 
Entropy - the lower score, the better split. Perfect split entropy=0.
Hellinger - the higher score, the better split. Perfect split helli=sqrt(2). 

This is why I added "1-" and specifically 1 because it is the highest value a class specific part in hellinger formula can get

All this is per my understanding of sklearn tree features split mechanism
Please correct me if I'm wrong

Reply:
My understanding of the split mechanism is rather shallow.
But upon looking into it, I see what you mean. You're right, we have to minimize the node `impurity` (smaller values are better).
Another alternative to accomplish the same is to flip the sign of the impurity in the final step. Instead of `hellinger_left += 1 - pow((count_k1  - count_k2), 2)` we could use something like `hellinger_left += -pow((count_k1  - count_k2), 2)`.
What do you think?

Reply:
I tried '-score' instead of '1-score', it should have the same effect, but it messes up the feature importances
so in order to make this change I'll need to get deeper into the feature importance mechanism in sklearn
would it be possible stay with current implementation?